src/kaboom.ts

```diff
@@ -5467,11 +5477,335 @@ export default (gopt: KaboomOpt = {}): KaboomCtx => {
 
 	}
 
-	function obstacle(): ObstacleComp {
+	function obstacle(isObstacle: boolean = true): ObstacleComp {
+		let _isObstacle = isObstacle
 		return {
 			id: "obstacle",
 			require: [ "tile" ],
-			isObstacle: true,
+			setObstacle(this: GameObj<ObstacleComp | TileComp>, isObstacle: boolean) {
+				if (_isObstacle !== isObstacle) {
+					_isObstacle = isObstacle
+					this.getLevel().invalidateNavigationMap()
+				}
+			},
+			isObstacle() {
+				return _isObstacle
+			},
+		}
+	}
+
+	function cost(cost: number = 1) : CostComp {
+		return {
+			id: "cost",
+			require: ["tile"],
+			cost: cost,
+		}
+	}
+
+	function edges(mask: EdgeMask | string[] = EdgeMask.All) : EdgesComp {
+		const lookup = {
+			"left": EdgeMask.Left,
+			"top": EdgeMask.Top,
+			"right": EdgeMask.Right,
+			"bottom": EdgeMask.Bottom,
+		}
+
+		const _mask: EdgeMask = Array.isArray(mask) ?
+			mask.map(s => lookup[s] || 0).reduce((mask, dir) => mask | dir, 0) :
+			mask
+
+		return {
+			id: "edges",
+			require: ["tile"],
+			mask: _mask,
+		}
+	}
+
+	function pathfinding() : PathfindingComp {
+		let _costMap: number[] | undefined
+		let _edgeMap: number[] | undefined
+		let _connectivityMap: number[] | undefined
+	
+		return {
+			id: "pathfinding",
+			require: ["level"],
+			add(this: GameObj<LevelComp | PathfindingComp>) {
+				this.onNavigationMapInvalid(() => {
+					this.invalidateNavigationMap()
+					this.trigger("navigation_map_changed")
+				})
+			},
+	
+			invalidateNavigationMap() {
+				_costMap = null
+				_edgeMap = null
+				_connectivityMap = null
+			},
+	
+			onNavigationMapChanged(this: GameObj<LevelComp | PathfindingComp>, cb: () => void) {
+				return this.on("navigation_map_changed", cb)
+			},
+	
+			getTilePath(this: GameObj<LevelComp | PathfindingComp>, from: Vec2, to: Vec2, diagonals: boolean) {
+				if (!_costMap) {
+					this._createCostMap()
+				}
+				if (!_edgeMap) {
+					this._createEdgeMap()
+				}
+				if (!_connectivityMap) {
+					this._createConnectivityMap()
+				}
+	
+				// Tiles are outside the grid
+				const numColumns = this.numColumns()
+				const numRows = this.numRows()
+				if (from.x < 0 || from.x >= numColumns ||
+					from.y < 0 || from.y >= numRows) {
+					return null
+				}
+				if (to.x < 0 || to.x >= numColumns ||
+					to.y < 0 || to.y >= numRows) {
+					return null
+				}
+	
+				const start = this._tile2Hash(from)
+				const goal = this._tile2Hash(to)
+	
+				// Tiles are not accessible
+				// If we test the start tile, we may get stuck
+				/*if (_costMap[start] === Infinity) {
+					return null
+				}*/
+				if (_costMap[goal] === Infinity) {
+					return null
+				}
```

Comment:
What about an an option to navigate to the closest non-obstacle tile, is that common?

Reply:
Changed it. Didn't know we used get/set from reading the Kaboom2000 code. I like get/set better for most things.

Reply:
I'll look into it. With the raycasting component you can look for the closest accessible (from your position) tile on a straight line towards the goal. But A* pathfinding will just hang if you feed it inaccessible tiles, therefore you check the connectivity map first. You could use breadth first search and then choose one of the closest tiles, but it might be slow.

Reply:
Definitely not a required feature for first launch, don't have a good idea on how to choose the closest point now either. Can also pick one from the neighbors of the goal tile, that's also has the closest distance to the start tile?

Reply:
I think we can use the connectivity map to look for the point in the current region which is closest to the target. However, that point may not always make sense. If you have a wall with a closed door, and the door has the possibility to connect the current region with a region which is closer to the target, it makes more sense to walk and stop in front of the door than towards a wall which won't budge.
So you would need to know potential connections somehow, obstacles which have the potential to lose their obstacle status and let you pass.
But if you know which obstacles can be unblocked, the best strategy would be to calculate the path as if those obstacles don't exist. This way the agent won't walk towards the door of a room which has no other exits, which would happen if it just looked for the closest position.

```
|---------|
|  x      |
|----|--d-|
|    |    |
|--d-|--d-|
| a       |
|------d--|
```